### PR TITLE
Allow zero-size read/write on pal stream

### PR DESCRIPTION
**Pull request by yamahata**
## Affected components

- [ ] README and global configuration
- [x] Linux PAL
- [ ] SGX PAL
- [ ] FreeBSD PAL
- [x] Common PAL code
- [ ] Library OS (i.e., SHIM), including GLIBC

## Description of the changes 
Allow zero size read/write on PAL stream

## How to test this PR? 

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1106)

### CODE REVIEW

File: LibOS/shim/src/fs/pipe/fs.c [hunk 1]
```diff
@@ -40,28 +40,22 @@
...
-    if (!count)
-        return 0;
-
     PAL_NUM bytes = DkStreamRead(hdl->pal_handle, 0, count, buf, NULL, 0);
 
     if (!bytes)
         return -PAL_ERRNO;
 
-    assert((ssize_t)bytes > 0);
+    assert(count == 0 || (ssize_t)bytes > 0);
```
Comment (sudharkrish) [inline LibOS/shim/src/fs/pipe/fs.c]: If we are removing !count check, before calling DkStreamRead, then we end up adding it part of assert. This makes an assumption that host kernel would have returned error(negative), for count == 0. I think we should either leave this, as-it-exists, (OR) think if we really need any assert here, why not let the user-application handle the error.

File: LibOS/shim/src/fs/pipe/fs.c [hunk 2]
```diff
@@ -40,28 +40,22 @@
...
-    if (!count)
-        return 0;
-
     PAL_NUM bytes = DkStreamWrite(hdl->pal_handle, 0, count, (void*)buf, NULL);
 
     if (!bytes)
         return -PAL_ERRNO;
 
-    assert((ssize_t)bytes > 0);
+    assert(count == 0 || (ssize_t)bytes > 0);
```
Comment (sudharkrish) [inline LibOS/shim/src/fs/pipe/fs.c]: Ditto.

File: LibOS/shim/src/fs/socket/fs.c [hunk 3]
```diff
@@ -81,7 +78,7 @@ static ssize_t socket_read(struct shim_handle* hdl, void* buf, size_t count) {
             }
         }
 
-    assert((ssize_t)bytes > 0);
+    assert(count == 0 || (ssize_t)bytes > 0);
```
Comment (sudharkrish) [inline LibOS/shim/src/fs/socket/fs.c]: Ditto.

File: LibOS/shim/src/fs/socket/fs.c [hunk 4]
```diff
@@ -127,7 +121,7 @@ static ssize_t socket_write(struct shim_handle* hdl, const void* buf, size_t cou
         return -err;
     }
 
-    assert((ssize_t)bytes > 0);
+    assert(count == 0 || (ssize_t)bytes > 0);
```
Comment (sudharkrish) [inline LibOS/shim/src/fs/socket/fs.c]: Ditto.

**Comment by yamahata** [comment]
Jenkins, retest this please.

**Comment by dimakuv** [comment]
Jenkins, retest Jenkins-Debug please

**Comment by dimakuv** [comment]
Jenkins, retest Jenkins-Debug-18.04 please

**Comment by dimakuv** [comment]
I had to update this PR slightly since the already-merged PR on eventfd introduced a new kind of streams (eventfd streams), and that code had to be updated similarly to how pipe code was updated by this PR.

**Pull request merged**